util/noeintr_close_shared.c

```diff
@@ -0,0 +1,212 @@
+#include <sys/socket.h>
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "entropy.h"
+#include "warnp.h"
+
+#include "noeintr.h"
+
+/* Sanity check compiling. */
+#ifndef NOEINTR_CLOSE_PTHREAD
+#error "Do not use this file directly; compile _normal.c or _pthread.c instead"
+#endif
+
+/* Handle multithreading (if applicable). */
+#if NOEINTR_CLOSE_PTHREAD
+#include <pthread.h>
+
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+#define MAGIC_COOKIE_LEN 32
+
+static int initialized = 0;
+static int s[2] = {-1, -1};
+static uint8_t cookie_write[MAGIC_COOKIE_LEN];
+
+#ifdef NOEINTR_CLOSE_TESTING
+/*
+ * POSIX is silent about the state of fd if close(2) gets EINTR.  For internal
+ * testing, this function randomly selects between the return values which are
+ * permitted by the standard.  For more info, see:
+ * https://www.daemonology.net/blog/2011-12-17-POSIX-close-is-broken.html
+ * https://www.austingroupbugs.net/view.php?id=529
+ */
+static int
+evil_close(int fd)
+{
+
+	switch (random() % 3) {
+	case 0:
+		/* Normal close(). */
+		return (close(fd));
+	case 1:
+		/* No close(), then return EINTR. */
+		errno = EINTR;
+		return (-1);
+	case 2:
+		/* Normal close(), then return EINTR. */
+		close(fd);
+		errno = EINTR;
+		return (-1);
+	}
+
+	/* Should never arrive here. */
+	warn0("evil_close() should not be here");
+	exit(1);
+}
+#endif
+
+static void
+noeintr_close_atexit(void)
+{
+
+	/* Attempt to close s[0] and s[1] using the normal close(). */
+	if (close(s[0]))
+		warnp("close");
+	if (close(s[1]))
+		warnp("close");
+}
+
+static int
+noeintr_close_init(void)
+{
+
+	/* Create a socket pair. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, s)) {
+		warnp("socketpair");
+		goto err0;
+	}
+
+	/* Set up cleanup function. */
+	if (atexit(noeintr_close_atexit)) {
+		warnp("atexit");
+		goto err1;
+	}
+
+	/* Generate a random cookie. */
+	if (entropy_read(cookie_write, MAGIC_COOKIE_LEN)) {
```

Comment:
This needs to be split into its component functions, yes?  Or were you planning on doing that when you switch `entropy_read` over to using `noeintr_close`?